src/sage/schemes/hyperelliptic_curves/jacobian_endomorphisms.py

```diff
@@ -0,0 +1,465 @@
+
+
+def satisfies_coefficient_condition(g,p):
+    """
+    This is the coefficient condition in the definition of Omega_K'
+    on page 912 of the published version of paper.
+
+    EXAMPLES::
+
+        sage: from sage.schemes.hyperelliptic_curves.jacobian_endomorphisms import satisfies_coefficient_condition
+        sage: R.<x> = ZZ[]
+        sage: f = x^4 + x^3 + 17*x^2 + 5*x
+        sage: satisfies_coefficient_condition(f,17)
+        False
+        sage: f = x^4 + x^3 + 17*x^2 + 23*x + 23^2
+        sage: satisfies_coefficient_condition(f,23)
+        True
+    """
+
+    if g[0] != p**2:
+        return False
+    if g[3]*p != g[1]:
+        return False
+    if g[2]%p == 0:
+        return False
+    return True
+
+
+def get_is_geom_irred(f,C,bad_primes,B=200):
+    """
+    Implements Algorithm 4.10 from the paper.
+
+    An additional optimisation comes from Part (2) of Theorem 4.8, from
+    which we can conclude that the endomorphism ring is absolutely simple.
+
+    INPUT:
+
+    - ``f`` -- a polynomial defining the hyperelliptic curve.
+
+    - ``C`` -- the hyperelliptic curve.
+
+    - ``bad_primes`` -- the list of odd primes of bad reduction.
+
+    - ``B`` -- the bound which appears in the statement of the algorithm
+    from Lombardo's paper.
+
+    OUTPUT:
+
+        (bool1, bool2): Pair of booleans. `bool1` indicates if the
+        absolute endomorphism algebra is a field; `bool2` indicates if the
+        absolute endomorphism algebra is the field of rational numbers.
+
+    EXAMPLES::
+
+    This is LMFDB curve 940693.a.960693.1
+
+        sage: from sage.schemes.hyperelliptic_curves.jacobian_endomorphisms import get_is_geom_irred
+        sage: R.<x> = QQ[]
+        sage: f = 4*x^6 - 12*x^5 + 20*x^3 - 8*x^2 - 4*x + 1
+        sage: C = HyperellipticCurve(f)
+        sage: get_is_geom_irred(f,C,[13,269])
+        (False, False)
+
+    This is LMFDB curve 3125.a.3125.1
+
+        sage: f = 4*x^5 + 1
+        sage: C = HyperellipticCurve(f)
+        sage: get_is_geom_irred(f,C,[5])
+        (True, False)
+
+    This is LMFDB curve 277.a.277.2
+
+        sage: f = 4*x^6 - 36*x^4 + 56*x^3 - 76*x^2 + 44*x - 23
+        sage: C = HyperellipticCurve(f)
+        sage: get_is_geom_irred(f,C,[277])
+        (True, True)
+
+    """
+
+    if C.has_odd_degree_model():
+        C_odd = C.odd_degree_model()
+        f_odd, h_odd = C_odd.hyperelliptic_polynomials()
+        # if f was odd to begin with, then f_odd = f
+        assert f_odd.degree() == 5
+
+        if (4*f_odd + h_odd**2).degree() == 5:
+            f_new = 4*f_odd + h_odd**2
+            if f_new.is_irreducible():
+                # i.e. the Jacobian is absolutely simple
+                f_disc_odd_prime_exponents = [v for _,v in f_new.discriminant().prime_to_S_part([ZZ(2)]).factor()]
+                if 1 in f_disc_odd_prime_exponents:
+                    return (True, True)  # Theorem 4.8 (2)
+                return (True, False)  # Algorithm 4.10 Step 1
```

Comment:
+ Proposition 4.7, as Gal(f) contains an element of order 5, thus there is a good prime p, such that f(x) is also irreducible, and thus we discard the QM option.

Reply:
Done